src/mapping/query/function/slice.rs

```diff
@@ -0,0 +1,269 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct SliceFn {
+    query: Box<dyn Function>,
+    start: Box<dyn Function>,
+    end: Option<Box<dyn Function>>,
+}
+
+impl SliceFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        query: Box<dyn Function>,
+        start: isize,
+        end: Option<isize>,
+    ) -> Self {
+        let start = Box::new(Literal::from(Value::from(start as i64)));
+        let end = end.map(|i| Box::new(Literal::from(Value::from(i as i64))) as _);
+
+        Self { query, start, end }
+    }
+}
+
+impl Function for SliceFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let mut start = match self.start.execute(ctx)? {
+            Value::Integer(v) => v,
+            v => unexpected_type!(v),
+        };
+
+        let end = &self
+            .end
+            .as_ref()
+            .map(|v| v.execute(ctx))
+            .transpose()?
+            .map(|v| match v {
+                Value::Integer(v) => v,
+                v => unexpected_type!(v),
+            });
+
+        let mut range = |len: i64| {
+            if start < 0 {
+                start += len;
+            }
+
+            let mut end = match end {
+                Some(end) => *end,
+                None => len,
+            };
+
+            if end < 0 {
+                end += len;
+            }
+
+            if start < 0 || start >= len {
+                return Err(format!(
+                    "'start' must be between '{}' and '{}'",
+                    -len,
+                    len - 1
+                ));
+            }
+
+            if end > len {
```

Comment:
I think we should allow `end` to be greater than `len`.

I imagine the intention of this function could be to ensure the text does not go over some limit, in which case shorter text is acceptable, but too long text gets truncated so it can fit. Presumably the script could cater for this with something like `slice(.foo, 0, min(5, len(.foo)))` but that's just extra processing overhead that doesn't have to be necessary.

Reply:
That's a fair point. We can relax the overflow errors 👍

(for text truncating, I'd say [`truncate`](https://github.com/timberio/vector/issues/3751) would be the preferred function)